DeadObjectException with SNAPSHOT 3.0.1

We hit a DeadObjectException using the following SNAPSHOT 
> com.mapzen.android:lost:3.0.1-20170607.212149-4

Sadly we don't have a reprodicible use-case. Will update the ticket if we have more information. 

```
9: 32: 05.801 10677 - 10677 / com.mapbox.mapboxsdk.testapp: lost E / AndroidRuntime: FATAL EXCEPTION: main
Process: com.mapbox.mapboxsdk.testapp: lost, PID: 10677
java.lang.RuntimeException: android.os.DeadObjectException
at com.mapzen.android.lost.internal.FusedLocationProviderServiceDelegate.reportLocation(FusedLocationProviderServiceDelegate.java: 78)
at com.mapzen.android.lost.internal.FusionEngine.onLocationChanged(FusionEngine.java: 196)
at android.location.LocationManager$ListenerTransport._handleMessage(LocationManager.java: 281)
at android.location.LocationManager$ListenerTransport.access$000(LocationManager.java: 210)
at android.location.LocationManager$ListenerTransport$2.handleMessage(LocationManager.java: 233)
at android.os.Handler.dispatchMessage(Handler.java: 102)
at android.os.Looper.loop(Looper.java: 135)
at android.app.ActivityThread.main(ActivityThread.java: 5312)
at java.lang.reflect.Method.invoke(Native Method)
at java.lang.reflect.Method.invoke(Method.java: 372)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java: 901)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java: 696)
Caused by: android.os.DeadObjectException
at android.os.BinderProxy.transactNative(Native Method)
at android.os.BinderProxy.transact(Binder.java: 496)
at com.mapzen.android.lost.internal.IFusedLocationProviderCallback$Stub$Proxy.onLocationChanged(IFusedLocationProviderCallback.java: 104)
at com.mapzen.android.lost.internal.FusedLocationProviderServiceDelegate.reportLocation(FusedLocationProviderServiceDelegate.java: 76)
at com.mapzen.android.lost.internal.FusionEngine.onLocationChanged(FusionEngine.java: 196)
at android.location.LocationManager$ListenerTransport._handleMessage(LocationManager.java: 281)
at android.location.LocationManager$ListenerTransport.access$000(LocationManager.java: 210)
at android.location.LocationManager$ListenerTransport$2.handleMessage(LocationManager.java: 233)
at android.os.Handler.dispatchMessage(Handler.java: 102)
at android.os.Looper.loop(Looper.java: 135)
at android.app.ActivityThread.main(ActivityThread.java: 5312)
at java.lang.reflect.Method.invoke(Native Method)
at java.lang.reflect.Method.invoke(Method.java: 372)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java: 901)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java: 696)
```

cc @sarahlensing @Guardiola31337 @zugaldia

Apart from the above `RuntimeException` caused by `DeadObjectException`, for Android 5.1 we're also getting a `RuntimeException` caused by `TransactionTooLargeException`
`java.lang.RuntimeException: 
  at com.mapzen.android.lost.internal.FusedLocationProviderServiceDelegate.reportLocation(FusedLocationProviderServiceDelegate.java:78)
  at com.mapzen.android.lost.internal.FusionEngine.onLocationChanged(FusionEngine.java:196)
  at android.location.LocationManager$ListenerTransport._handleMessage(LocationManager.java:286)
  at android.location.LocationManager$ListenerTransport.access$000(LocationManager.java:215)
  at android.location.LocationManager$ListenerTransport$2.handleMessage(LocationManager.java:238)
  at android.os.Handler.dispatchMessage(Handler.java:102)
  at android.os.Looper.loop(Looper.java:145)
  at android.app.ActivityThread.main(ActivityThread.java:6934)
  at java.lang.reflect.Method.invoke(Native Method:0)
  at java.lang.reflect.Method.invoke(Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1404)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1199)
Caused by: android.os.TransactionTooLargeException: 
  at android.os.BinderProxy.transactNative(Native Method:0)
  at android.os.BinderProxy.transact(Binder.java:496)
  at com.mapzen.android.lost.internal.IFusedLocationProviderCallback$Stub$Proxy.onLocationChanged(IFusedLocationProviderCallback.java:104)
  at com.mapzen.android.lost.internal.FusedLocationProviderServiceDelegate.reportLocation(FusedLocationProviderServiceDelegate.java:76)`

Also seeing a similar issue on a Nexus 6P and Android 7.1.2. It happens when a route navigation is active and I send the app to the background (press the home button for example). 

`java.lang.RuntimeException: android.os.DeadObjectException
	at com.mapzen.android.lost.internal.FusedLocationProviderServiceDelegate.reportLocation(FusedLocationProviderServiceDelegate.java:78)
	at com.mapzen.android.lost.internal.FusionEngine.onLocationChanged(FusionEngine.java:191)
	at android.location.LocationManager$ListenerTransport._handleMessage(LocationManager.java:297)
	at android.location.LocationManager$ListenerTransport.-wrap0(LocationManager.java)
	at android.location.LocationManager$ListenerTransport$2.handleMessage(LocationManager.java:249)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6121)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
Caused by: android.os.DeadObjectException
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(Binder.java:615)
	at com.mapzen.android.lost.internal.IFusedLocationProviderCallback$Stub$Proxy.onLocationChanged(IFusedLocationProviderCallback.java:104)
	at com.mapzen.android.lost.internal.FusedLocationProviderServiceDelegate.reportLocation(FusedLocationProviderServiceDelegate.java:76)`

--- a/lost/src/main/java/com/mapzen/android/lost/api/FusedLocationProviderApi.java
+++ b/lost/src/main/java/com/mapzen/android/lost/api/FusedLocationProviderApi.java
@@ -37,6 +37,8 @@ public interface FusedLocationProviderApi {
    * Network location is enabled and the {@link android.location.LocationManager} has a last known
    * Network location
    *
+   * This method can return null if there is a problem getting the availability.
+   *
    * @param client The client to return availability for.
    * @return The availability of location data.
    * @throws IllegalStateException if the client is not connected at the time of this call.

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
@@ -20,6 +20,7 @@ import android.os.Handler;
 import android.os.IBinder;
 import android.os.Looper;
 import android.os.RemoteException;
+import android.util.Log;
 
 import java.util.List;
 import java.util.Map;
@@ -31,6 +32,7 @@ import java.util.Set;
 public class FusedLocationProviderApiImpl extends ApiImpl
     implements FusedLocationProviderApi, EventCallbacks, ServiceConnection {
 
+  private static final String TAG = FusedLocationProviderApiImpl.class.getSimpleName();
   private Context context;
   private FusedLocationServiceConnectionManager serviceConnectionManager;
   private FusedLocationServiceCallbackManager serviceCallbackManager;
@@ -123,19 +125,25 @@ public class FusedLocationProviderApiImpl extends ApiImpl
 
   @Override public Location getLastLocation(LostApiClient client) {
     throwIfNotConnected(client);
+    Location location = null;
     try {
-      return service.getLastLocation();
+      location = service.getLastLocation();
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to get last Location " + e.getStackTrace());
+    } finally {
+      return location;
     }
   }
 
   @Override public LocationAvailability getLocationAvailability(LostApiClient client) {
     throwIfNotConnected(client);
+    LocationAvailability availability = null;
     try {
-      return service.getLocationAvailability();
+      availability = service.getLocationAvailability();
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to get LocationAvailability " + e.getStackTrace());
+    } finally {
+      return availability;
     }
   }
 
@@ -179,7 +187,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.requestLocationUpdates(request);
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to request location updates " + e.getStackTrace());
     }
   }
 
@@ -190,7 +198,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.removeLocationUpdates(requests);
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to remove location updates " + e.getStackTrace());
     }
   }
 
@@ -244,7 +252,8 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.setMockMode(isMockMode);
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      String mode = isMockMode ? "enabled" : "disabled";
+      Log.e(TAG, "Error occurred trying to set mock mode " + mode + " " + e.getStackTrace());
     }
     return new SimplePendingResult(true);
   }
@@ -255,7 +264,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.setMockLocation(mockLocation);
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to set mock location " + e.getStackTrace());
     }
     return new SimplePendingResult(true);
   }
@@ -266,7 +275,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.setMockTrace(path, filename);
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to set mock trace " + e.getStackTrace());
     }
     return new SimplePendingResult(true);
   }
@@ -288,7 +297,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
       try {
         service.init(remoteCallback);
       } catch (RemoteException e) {
-        throw new RuntimeException(e);
+        Log.e(TAG, "Error occurred trying to register remote callback " + e.getStackTrace());
       }
     }
   }
@@ -298,7 +307,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
       try {
         service.init(null);
       } catch (RemoteException e) {
-        throw new RuntimeException(e);
+        Log.e(TAG, "Error occurred trying to unregister remote callback " + e.getStackTrace());
       }
     }
   }

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
@@ -129,7 +129,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       location = service.getLastLocation();
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to get last Location " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to get last Location", e);
     } finally {
       return location;
     }
@@ -141,7 +141,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       availability = service.getLocationAvailability();
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to get LocationAvailability " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to get LocationAvailability", e);
     } finally {
       return availability;
     }
@@ -187,7 +187,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.requestLocationUpdates(request);
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to request location updates " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to request location updates", e);
     }
   }
 
@@ -198,7 +198,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.removeLocationUpdates(requests);
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to remove location updates " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to remove location updates", e);
     }
   }
 
@@ -253,7 +253,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
       service.setMockMode(isMockMode);
     } catch (RemoteException e) {
       String mode = isMockMode ? "enabled" : "disabled";
-      Log.e(TAG, "Error occurred trying to set mock mode " + mode + " " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to set mock mode " + mode, e);
     }
     return new SimplePendingResult(true);
   }
@@ -264,7 +264,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.setMockLocation(mockLocation);
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to set mock location " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to set mock location", e);
     }
     return new SimplePendingResult(true);
   }
@@ -275,7 +275,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
     try {
       service.setMockTrace(path, filename);
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to set mock trace " + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to set mock trace", e);
     }
     return new SimplePendingResult(true);
   }
@@ -297,7 +297,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
       try {
         service.init(remoteCallback);
       } catch (RemoteException e) {
-        Log.e(TAG, "Error occurred trying to register remote callback " + e.getStackTrace());
+        Log.e(TAG, "Error occurred trying to register remote callback", e);
       }
     }
   }
@@ -307,7 +307,7 @@ public class FusedLocationProviderApiImpl extends ApiImpl
       try {
         service.init(null);
       } catch (RemoteException e) {
-        Log.e(TAG, "Error occurred trying to unregister remote callback " + e.getStackTrace());
+        Log.e(TAG, "Error occurred trying to unregister remote callback", e);
       }
     }
   }

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceDelegate.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceDelegate.java
@@ -10,6 +10,7 @@ import android.os.Bundle;
 import android.os.Looper;
 import android.os.RemoteException;
 import android.support.annotation.RequiresPermission;
+import android.util.Log;
 
 import java.io.File;
 import java.util.List;
@@ -19,6 +20,7 @@ import static android.Manifest.permission.ACCESS_FINE_LOCATION;
 
 public class FusedLocationProviderServiceDelegate implements LocationEngine.Callback {
 
+  private static final String TAG = FusedLocationProviderServiceDelegate.class.getSimpleName();
   private Context context;
 
   private boolean mockMode;
@@ -76,7 +78,8 @@ public class FusedLocationProviderServiceDelegate implements LocationEngine.Call
       try {
         callback.onLocationChanged(location);
       } catch (RemoteException e) {
-        throw new RuntimeException(e);
+        Log.e(TAG, "Error occurred trying to report a new Location. "
+            + e.getStackTrace().toString());
       }
     }
   }
@@ -130,7 +133,8 @@ public class FusedLocationProviderServiceDelegate implements LocationEngine.Call
       try {
         callback.onLocationAvailabilityChanged(availability);
       } catch (RemoteException e) {
-        throw new RuntimeException(e);
+        Log.e(TAG, "Error occurred trying to report a new LocationAvailability. "
+            + e.getStackTrace().toString());
       }
     }
   }

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceDelegate.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceDelegate.java
@@ -78,8 +78,7 @@ public class FusedLocationProviderServiceDelegate implements LocationEngine.Call
       try {
         callback.onLocationChanged(location);
       } catch (RemoteException e) {
-        Log.e(TAG, "Error occurred trying to report a new Location. "
-            + e.getStackTrace().toString());
+        Log.e(TAG, "Error occurred trying to report a new Location", e);
       }
     }
   }
@@ -133,8 +132,7 @@ public class FusedLocationProviderServiceDelegate implements LocationEngine.Call
       try {
         callback.onLocationAvailabilityChanged(availability);
       } catch (RemoteException e) {
-        Log.e(TAG, "Error occurred trying to report a new LocationAvailability. "
-            + e.getStackTrace().toString());
+        Log.e(TAG, "Error occurred trying to report a new LocationAvailability", e);
       }
     }
   }

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceCallbackManager.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceCallbackManager.java
@@ -6,8 +6,11 @@ import com.mapzen.android.lost.api.LocationResult;
 import android.content.Context;
 import android.location.Location;
 import android.os.RemoteException;
+import android.util.Log;
 
 import java.util.ArrayList;
+
+import static android.content.ContentValues.TAG;
 
 /**
  * Handles callbacks received in {@link FusedLocationProviderApiImpl} from
@@ -29,11 +32,11 @@ public class FusedLocationServiceCallbackManager {
 
     ReportedChanges changes = clientManager.reportLocationChanged(location);
 
-    LocationAvailability availability;
+    LocationAvailability availability = null;
     try {
       availability = service.getLocationAvailability();
     } catch (RemoteException e) {
-      throw new RuntimeException(e);
+      Log.e(TAG, "Error occurred trying to get LocationAvailability" + e.getStackTrace());
     }
 
     ArrayList<Location> locations = new ArrayList<>();

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceCallbackManager.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceCallbackManager.java
@@ -36,7 +36,7 @@ public class FusedLocationServiceCallbackManager {
     try {
       availability = service.getLocationAvailability();
     } catch (RemoteException e) {
-      Log.e(TAG, "Error occurred trying to get LocationAvailability" + e.getStackTrace());
+      Log.e(TAG, "Error occurred trying to get LocationAvailability", e);
     }
 
     ArrayList<Location> locations = new ArrayList<>();

